django_mongodb/compiler.py

```diff
@@ -473,12 +451,47 @@ def get_project_fields(self, columns=None, ordering=None):
                 if self.query.alias_refcount[alias] and self.collection_name != alias:
                     fields[alias] = 1
             # Add order_by() fields.
-            for column, _ in ordering or []:
-                foreign_table = column.split(".", 1)[0] if "." in column else None
-                if column not in fields and foreign_table not in fields:
-                    fields[column] = 1
+            for alias, expression in ordering or []:
+                nested_entity = alias.split(".", 1)[0] if "." in alias else None
+                if alias not in fields and nested_entity not in fields:
+                    fields[alias] = expression.as_mql(self, self.connection)
         return fields
 
+    def _get_ordering(self):
+        """
+        Process the query's OrderBy objects and return:
+        - A tuple of ('field_name': Col/Expression, ...)
+        - A bson.SON mapping to pass to $sort.
+        - A tuple of ('field_name': Expression, ...) for expressions that need
+          to be added to extra_fields.
+        """
+        fields = {}
+        sort_ordering = SON()
+        extra_fields = {}
+        idx = itertools.count(start=1)
+        for order in self.order_by_expressions or []:
+            if isinstance(order.expression, Ref):
+                # TODO change?
+                # field_name = order.expression.as_mql(self, self.connection).removeprefix("$")
+                # extra_fields[field_name] = order.expression
+                field_name = order.expression.refs
+            elif isinstance(order.expression, Col):
+                field_name = order.expression.as_mql(self, self.connection).removeprefix("$")
+                fields[field_name] = order.expression
+            else:
+                # The expression must be added to extra_fields with an alias.
+                field_name = f"__order{next(idx)}"
+                fields[field_name] = order.expression
```

Comment:
the reason is:
The `refs` are actually references to specific fields selected in a query. These references compile to a position number within the `as_sql` method. To address this, I decided to move the `refs` to the post-project fields. Additionally, some of these `refs` are calculated during the project stage (such as annotations), so the `refs` are processed after the project stage. But the other projected fields that aren't `refs` have all the info to make the full expression.

Btw, 'refs' also has all the info to be compiled well, like an annotation with name (something that is not  `__annotation` but isn't a column), it has the reference but also the full expression with it. So if we want to delete the extra field we could, the downside will be a field will be calculated twice or more.

Reply:
Even with the explanation, it's still difficult to understand. Now in this method, `extra_fields` is only used for nulls first/last, but then outside this method, all `ordering_fields` are added to `extra_fields` when `columns is None` (I guess because they won't be added to `$project` in that case).